esp-hal/src/gpio/mod.rs

```diff
@@ -930,8 +930,11 @@ fn handle_pin_interrupts(user_handler: fn()) {
             intr_bits -= 1 << pin_pos;
 
             let pin_nr = pin_pos as u8 + bank.offset();
-            asynch::PIN_WAKERS[pin_nr as usize].wake();
-            set_int_enable(pin_nr, Some(0), 0, false);
+
+            crate::interrupt::free(|| {
+                asynch::PIN_WAKERS[pin_nr as usize].wake();
+                set_int_enable(pin_nr, Some(0), 0, false);
+            });
```

Comment:
Did you change your mind? This doesn't match the PR title.
I was expecting you to swap the order of waking and interrupt clearing.

Reply:
Well, it depends how you look at it, the task is waken immediately, or effectively at the end of the interrupt-free section. I'm not sure that swapping the order would be correct.

Reply:
The infinite loop is caused because the waker is woken before the interrupts are cleared right? Swapping the order seems correct to me in that case. "Don't wake up the task until the pin's interrupt has been completely serviced"

Reply:
I guess it might be safe enough, but I'm not confident enough.